File: JavaScript/BeautifulArrangement/BeautifulArrangement.js
/**
 * @param {number} N
 * @return {number}
 */
// (172ms)
var countArrangement = function(N) {
    var helper = function(used, N, now){
        if(now < 2) {
            ans++;
            return;
        }
        for(let i = 1; i <= N; i++) {
            if(used[i])
                continue;
            if(i % now == 0 || now % i == 0) {
                used[i] = true;
                helper(used, N, now - 1);
                used[i] = false;
            }
        }
    };
    var ans = 0;
    used = new Array(N + 1);
    used.fill(false);
    helper(used, N, N);
    return ans;
};

File: Python/BeautifulArrangement/BeautifulArrangement.py
class Solution(object):

    ans = 0
    # (182ms)
    def countArrangement(self, N):
        """
        :type N: int
        :rtype: int
        """
        def helper(used, N, now):
            if now < 2:
                self.ans += 1
                return
            for i in xrange(1, N + 1):
                if used[i]:
                    continue
                if i % now == 0 or now % i == 0:
                    used[i] = True
                    helper(used, N, now - 1)
                    used[i] = False

        used = [False] * (N + 1)
        helper(used, N, N)
        return self.ans
